File: mParticle-Apple-SDK/Data Model/MPSession.m
#import "MPSession.h"
#import "MPIConstants.h"
#import "MPPersistenceController.h"
@import mParticle_Apple_SDK_Swift;

@interface MPSession ()
@property (nonatomic, strong) MPSessionPRIVATE *implementation;
@end

@implementation MPSession

- (instancetype)init {
    NSTimeInterval now = [[NSDate date] timeIntervalSince1970];
    NSNumber *mpId = [MPPersistenceController_PRIVATE mpId];
    return [self initWithSessionId:0
                              UUID:[[NSUUID UUID] UUIDString]
                    backgroundTime:0.0
                         startTime:now
                           endTime:now
                        attributes:nil
             numberOfInterruptions:0
                      eventCounter:0
                       suspendTime:0
                            userId:mpId
                    sessionUserIds:[mpId stringValue]
                           appInfo:nil
                        deviceInfo:nil];
}

- (instancetype)initWithStartTime:(NSTimeInterval)timestamp userId:(NSNumber *)userId {
    return [self initWithStartTime:timestamp userId:userId uuid:nil];
}

- (instancetype)initWithStartTime:(NSTimeInterval)timestamp userId:(NSNumber *)userId uuid:(NSString *)uuid {
    NSString *uuidString = uuid ?: [[NSUUID UUID] UUIDString];
    return [self initWithSessionId:0
                              UUID:uuidString
                    backgroundTime:0.0
                         startTime:timestamp
                           endTime:timestamp
                        attributes:nil
             numberOfInterruptions:0
                      eventCounter:0
                       suspendTime:0
                            userId:userId
                    sessionUserIds:[userId stringValue]
                           appInfo:nil
                        deviceInfo:nil];
}

- (instancetype)initWithSessionId:(int64_t)sessionId
                             UUID:(NSString *)uuid
                   backgroundTime:(NSTimeInterval)backgroundTime
                        startTime:(NSTimeInterval)startTime
                          endTime:(NSTimeInterval)endTime
                       attributes:(NSMutableDictionary *)attributesDictionary
            numberOfInterruptions:(uint)numberOfInterruptions
                     eventCounter:(uint)eventCounter
                      suspendTime:(NSTimeInterval)suspendTime
                           userId:(NSNumber *)userId
                   sessionUserIds:(NSString *)sessionUserIds
                          appInfo:(nullable NSDictionary<NSString *,id> *)appInfo
                       deviceInfo:(nullable NSDictionary *)deviceInfo
{
    self = [super init];
    if (!self) {
        return nil;
    }

    _implementation = [[MPSessionPRIVATE alloc] initWithSessionId:sessionId
                                                             uuid:uuid
                                                   backgroundTime:backgroundTime
                                                        startTime:startTime
                                                          endTime:endTime
                                                       attributes:attributesDictionary
                                            numberOfInterruptions:numberOfInterruptions
                                                     eventCounter:eventCounter
                                                      suspendTime:suspendTime
                                                           userId:userId
                                                   sessionUserIds:sessionUserIds
                                                  applicationInfo:appInfo
                                                       deviceInfo:deviceInfo];
    _uuid = _implementation.uuid;
    return self;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"Session\n Id: %lld\n UUID: %@\n Background time: %.0f\n Foreground time: %.0f\n Start: %.0f\n End: %.0f\n Length: %.0f\n EventCounter: %d\n Persisted: %d\n Interruptions: %d\n Attributes: %@\n", self.sessionId, self.uuid, self.backgroundTime, self.foregroundTime, self.startTime, self.endTime, self.length, self.eventCounter, self.persisted, self.numberOfInterruptions, self.attributesDictionary];
}

- (BOOL)isEqual:(MPSession *)object {
    if (MPIsNull(object) || ![object isKindOfClass:[MPSession class]]) {
        return NO;
    }
    return [self.implementation isEqualToSession:object.implementation];
}

- (NSUInteger)hash {
    return (NSUInteger)self.implementation.hash;
}

- (id)copyWithZone:(NSZone *)zone {
    MPSession *copyObject = [[[self class] allocWithZone:zone] init];
    copyObject.implementation = [self.implementation copySession];
    copyObject.uuid = copyObject.implementation.uuid;
    return copyObject;
}

- (NSMutableDictionary *)attributesDictionary {
    return self.implementation.attributesDictionary;
}

- (void)setAttributesDictionary:(NSMutableDictionary *)attributesDictionary {
    self.implementation.attributesDictionary = attributesDictionary ?: [NSMutableDictionary dictionary];
}

- (NSTimeInterval)backgroundTime {
    return self.implementation.backgroundTime;
}

- (void)setBackgroundTime:(NSTimeInterval)backgroundTime {
    self.implementation.backgroundTime = backgroundTime;
}

- (NSTimeInterval)foregroundTime {
    return self.implementation.foregroundTime;
}

- (NSTimeInterval)startTime {
    return self.implementation.startTime;
}

- (void)setStartTime:(NSTimeInterval)startTime {
    self.implementation.startTime = startTime;
}

- (NSTimeInterval)endTime {
    return self.implementation.endTime;
}

- (void)setEndTime:(NSTimeInterval)endTime {
    [self.implementation applyEndTime:endTime];
}

- (NSTimeInterval)length {
    BOOL needsKVO = self.implementation.length == 0 && self.implementation.endTime > self.implementation.startTime;
    if (needsKVO) {
        [self willChangeValueForKey:@"length"];
    }
    NSTimeInterval length = [self.implementation resolveLength];
    if (needsKVO) {
        [self didChangeValueForKey:@"length"];
    }
    return length;
}

- (void)setLength:(NSTimeInterval)length {
    self.implementation.length = length;
}

- (NSTimeInterval)suspendTime {
    return self.implementation.suspendTime;
}

- (uint)eventCounter {
    return self.implementation.eventCounter;
}

- (uint)numberOfInterruptions {
    return self.implementation.numberOfInterruptions;
}

- (int64_t)sessionId {
    return self.implementation.sessionId;
}

- (void)setSessionId:(int64_t)sessionId {
    self.implementation.sessionId = sessionId;
}

- (BOOL)persisted {
    return self.implementation.persisted;
}

- (NSNumber *)userId {
    return self.implementation.userId;
}

- (void)setUserId:(NSNumber *)userId {
    self.implementation.userId = userId;
}

- (NSString *)sessionUserIds {
    return self.implementation.sessionUserIds;
}

- (void)setSessionUserIds:(NSString *)sessionUserIds {
    self.implementation.sessionUserIds = sessionUserIds;
}

- (NSDictionary<NSString *, id> *)appInfo {
    return self.implementation.applicationInfo;
}

- (void)setAppInfo:(NSDictionary<NSString *, id> *)appInfo {
    self.implementation.applicationInfo = appInfo;
}

- (NSDictionary *)deviceInfo {
    return self.implementation.deviceInfo;
}

- (void)setDeviceInfo:(NSDictionary *)deviceInfo {
    self.implementation.deviceInfo = deviceInfo;
}

- (NSString *)uuid {
    return self.implementation.uuid;
}

- (void)setUuid:(NSString *)uuid {
    self.implementation.uuid = uuid;
    _uuid = uuid;
}

- (void)incrementCounter {
    [self willChangeValueForKey:@"eventCounter"];
    [self.implementation incrementCounter];
    [self didChangeValueForKey:@"eventCounter"];
}

- (void)suspendSession {
    [self willChangeValueForKey:@"numberOfInterruptions"];
    [self willChangeValueForKey:@"suspendTime"];
    [self.implementation suspendSession];
    [self didChangeValueForKey:@"numberOfInterruptions"];
    [self didChangeValueForKey:@"suspendTime"];
}

@end

File: mParticle-Apple-SDK-Swift/Sources/DataModel/MPSession.swift
import Foundation

@objc(MPSession)
public final class MPSessionPRIVATE: NSObject, NSCopying {
    private let lock = NSRecursiveLock()
    private var storedSessionId: Int64
    private var storedUUID: String
    private var storedBackgroundTime: TimeInterval
    private var storedStartTime: TimeInterval
    private var storedEndTime: TimeInterval
    private var storedLength: TimeInterval
    private var storedEventCounter: UInt32
    private var storedNumberOfInterruptions: UInt32
    private var storedSuspendTime: TimeInterval
    private var storedUserId: NSNumber
    private var storedSessionUserIds: String
    private var storedAttributesDictionary: NSMutableDictionary
    private var storedAppInfo: NSDictionary?
    private var storedDeviceInfo: NSDictionary?

    @objc dynamic public var sessionId: Int64 {
        get { withLock { storedSessionId } }
        set { withLock { storedSessionId = newValue } }
    }

    @objc dynamic public var uuid: String {
        get { withLock { storedUUID } }
        set { withLock { storedUUID = newValue } }
    }

    @objc dynamic public var backgroundTime: TimeInterval {
        get { withLock { storedBackgroundTime } }
        set { withLock { storedBackgroundTime = newValue } }
    }

    @objc dynamic public var startTime: TimeInterval {
        get { withLock { storedStartTime } }
        set { withLock { storedStartTime = newValue } }
    }

    @objc dynamic public var endTime: TimeInterval {
        get { withLock { storedEndTime } }
        set { applyEndTime(newValue) }
    }

    @objc dynamic public var length: TimeInterval {
        get {
            withLock {
                if storedLength == 0, storedEndTime > storedStartTime {
                    willChangeValue(forKey: "length")
                    storedLength = storedEndTime - storedStartTime
                    didChangeValue(forKey: "length")
                }
                return storedLength
            }
        }
        set { withLock { storedLength = newValue } }
    }

    @objc dynamic public private(set) var eventCounter: UInt32 {
        get { withLock { storedEventCounter } }
        set { withLock { storedEventCounter = newValue } }
    }

    @objc dynamic public private(set) var numberOfInterruptions: UInt32 {
        get { withLock { storedNumberOfInterruptions } }
        set { withLock { storedNumberOfInterruptions = newValue } }
    }

    @objc dynamic public private(set) var suspendTime: TimeInterval {
        get { withLock { storedSuspendTime } }
        set { withLock { storedSuspendTime = newValue } }
    }

    @objc dynamic public var userId: NSNumber {
        get { withLock { storedUserId } }
        set { withLock { storedUserId = newValue } }
    }

    @objc dynamic public var sessionUserIds: String {
        get { withLock { storedSessionUserIds } }
        set { withLock { storedSessionUserIds = newValue } }
    }

    @objc dynamic public var attributesDictionary: NSMutableDictionary {
        get { withLock { storedAttributesDictionary } }
        set { withLock { storedAttributesDictionary = newValue } }
    }

    @objc dynamic public var appInfo: NSDictionary? {
        get { withLock { storedAppInfo } }
        set { withLock { storedAppInfo = newValue } }
    }

    @objc dynamic public var deviceInfo: NSDictionary? {
        get { withLock { storedDeviceInfo } }
        set { withLock { storedDeviceInfo = newValue } }
    }

    @objc public var persisted: Bool {
        withLock { storedSessionId != 0 }
    }

    @objc public var foregroundTime: TimeInterval {
        withLock { max(storedLength - storedBackgroundTime, 0) }
    }

    override public convenience init() {
        let now = Date().timeIntervalSince1970
        self.init(startTime: now, userId: MPUserDefaults.storedMpId())
    }

    @objc(initWithStartTime:userId:)
    public convenience init(startTime: TimeInterval, userId: NSNumber) {
        self.init(startTime: startTime, userId: userId, uuid: nil)
    }

    @objc(initWithStartTime:userId:uuid:)
    public convenience init(startTime: TimeInterval, userId: NSNumber, uuid: String?) {
        self.init(
            sessionId: 0,
            uuid: uuid ?? UUID().uuidString,
            backgroundTime: 0,
            startTime: startTime,
            endTime: startTime,
            attributes: nil,
            numberOfInterruptions: 0,
            eventCounter: 0,
            suspendTime: 0,
            userId: userId,
            sessionUserIds: userId.stringValue,
            applicationInfo: nil,
            deviceInfo: nil
        )
    }

    // swiftlint:disable line_length
    @objc(
        initWithSessionId:UUID:backgroundTime:startTime:endTime:attributes:numberOfInterruptions:eventCounter:suspendTime:userId:sessionUserIds:appInfo:deviceInfo:
    )
    // swiftlint:enable line_length
    public init(
        sessionId: Int64,
        uuid: String,
        backgroundTime: TimeInterval,
        startTime: TimeInterval,
        endTime: TimeInterval,
        attributes: NSMutableDictionary?,
        numberOfInterruptions: UInt32,
        eventCounter: UInt32,
        suspendTime: TimeInterval,
        userId: NSNumber,
        sessionUserIds: String,
        applicationInfo: NSDictionary?,
        deviceInfo: NSDictionary?
    ) {
        storedSessionId = sessionId
        storedUUID = uuid
        storedBackgroundTime = backgroundTime
        storedStartTime = startTime
        storedEndTime = endTime
        storedLength = endTime - startTime
        storedEventCounter = eventCounter
        storedNumberOfInterruptions = numberOfInterruptions
        storedSuspendTime = suspendTime
        storedUserId = userId
        storedSessionUserIds = sessionUserIds
        storedAttributesDictionary = attributes ?? NSMutableDictionary()
        storedAppInfo = applicationInfo
        storedDeviceInfo = deviceInfo
        super.init()
    }

    @objc public func incrementCounter() {
        withLock {
            willChangeValue(forKey: "eventCounter")
            storedEventCounter += 1
            didChangeValue(forKey: "eventCounter")
        }
    }

    @objc public func suspendSession() {
        withLock {
            willChangeValue(forKey: "numberOfInterruptions")
            willChangeValue(forKey: "suspendTime")
            storedNumberOfInterruptions += 1
            storedSuspendTime = Date().timeIntervalSince1970
            didChangeValue(forKey: "numberOfInterruptions")
            didChangeValue(forKey: "suspendTime")
        }
    }

    @objc public func copy(with _: NSZone? = nil) -> Any {
        withLock {
            MPSessionPRIVATE(
                sessionId: storedSessionId,
                uuid: storedUUID,
                backgroundTime: storedBackgroundTime,
                startTime: storedStartTime,
                endTime: storedEndTime,
                attributes: storedAttributesDictionary.mutableCopy() as? NSMutableDictionary,
                numberOfInterruptions: storedNumberOfInterruptions,
                eventCounter: storedEventCounter,
                suspendTime: storedSuspendTime,
                userId: storedUserId,
                sessionUserIds: storedSessionUserIds,
                applicationInfo: storedAppInfo,
                deviceInfo: storedDeviceInfo
            )
        }
    }

    override public func isEqual(_ object: Any?) -> Bool {
        guard let other = object as? MPSessionPRIVATE else { return false }
        return isEqual(toSession: other)
    }

    @objc(isEqualToSession:)
    public func isEqual(toSession other: MPSessionPRIVATE) -> Bool {
        sessionId == other.sessionId && eventCounter == other.eventCounter && uuid == other.uuid
    }

    override public var hash: Int {
        withLock {
            Int(truncatingIfNeeded: storedSessionId) ^ Int(storedEventCounter) ^ storedUUID.hashValue
        }
    }

    override public var description: String {
        withLock {
            "Session\n Id: \(storedSessionId)\n UUID: \(storedUUID)\n Background time: "
                + "\(String(format: "%.0f", storedBackgroundTime))\n Foreground time: "
                + "\(String(format: "%.0f", max(storedLength - storedBackgroundTime, 0)))\n Start: "
                + "\(String(format: "%.0f", storedStartTime))\n End: \(String(format: "%.0f", storedEndTime))"
                + "\n Length: \(String(format: "%.0f", storedLength))\n EventCounter: \(storedEventCounter)"
                + "\n Persisted: \(storedSessionId != 0 ? 1 : 0)\n Interruptions: \(storedNumberOfInterruptions)"
                + "\n Attributes: \(storedAttributesDictionary)\n"
        }
    }

    private func applyEndTime(_ proposedEndTime: TimeInterval) {
        withLock {
            if proposedEndTime > storedStartTime {
                storedEndTime = proposedEndTime
                storedLength = storedEndTime - storedStartTime
            } else if storedLength > 0 {
                storedEndTime = storedStartTime + storedLength
            } else {
                storedEndTime = storedStartTime
            }
        }
    }

    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}
